Make numeric_range an iterable

jferard opened the pull request:
The `numeric_range` function is meant to be

> An extension of the built-in ``range()`` function whose arguments can be any orderable numeric type.

But `range(...)` is an iterable, while `numeric_range(...)` is an iterator:

    >>> r = range(5)
    >>> list(r) + list(r)
    [0, 1, 2, 3, 4, 0, 1, 2, 3, 4]
    >>> from more_itertools import numeric_range
    >>> nr = numeric_range(5)
    >>> list(nr) + list(nr) # nr was exhausted
    [0, 1, 2, 3, 4]

I think that `numeric_range` should be an iterable. (Note that this semantic difference has some effects: `range` has an efficient implementation of `len`, `reversed`, `r[...]`, but those are features of an iterable, not an iterator. Hence we can't add these features to `numeric_range`.)

 This PR is a basic implementation. Main drawback: we need division and modulo. I don't know if there are numeric entities in Python that can't be divided (and would need a fallback).

bbayles commented on more_itertools/more.py:
```diff
@@ -1590,7 +1590,7 @@ def groupby_transform(iterable, keyfunc=None, valuefunc=None):
 
 
-def numeric_range(*args):
+class numeric_range:
```
Would it be possible inherit from `collections.abc.Iterable` and `collections.abc.Sequence`? `range()` objects are marked as instances of each. Could add a test for this too.

jferard replied:
Good idea, but inherit from  ̀collections.abc.Iterable` is unnecessary since a `Sequence` is already an `Iterable`. I will mark `numeric_range` as  ̀collections.abc.Sequence` and  ̀collections.abc.Hashable`.

jferard replied:
Done.

bbayles commented on more_itertools/more.py:
```diff
@@ -1634,35 +1634,148 @@ def numeric_range(*args):
+        elif argc == 3:
+            self._start, self._stop, self._step = args
+        else:
+            if argc == 0:
```
Could you just move this into the `if / elif` structure above?

jferard replied:
You mean something like:

```
    elif argc == 0:
        raise TypeError('numeric_range expected at least 1 argument, got {}'.format(argc)
    else:
        raise TypeError('numeric_range expected at most 3 arguments, got {}'.format(argc)
```
?

(I added "at least" to make the message clearer).

bbayles replied:
Yeah, exactly.

jferard replied:
Done.

bbayles commented on more_itertools/more.py:
```diff
@@ -1634,35 +1634,148 @@ def numeric_range(*args):
+        else:
+            return self._EMPTY_HASH
+
+    def __iter__(self):
```
I think this could be re-worked.

First, doing repeated `+=` operations is going to introduce accumulating floating point error.

Second, I think `__iter__` should return `self` and `__next__` should return the next `_get_by_index` value (meaning, keep track of `n` like [here](https://github.com/erikrose/more-itertools/blob/c11195abbc3ca7408d77ac051712162b455b2306/more_itertools/more.py#L1660-L1665)).

jferard replied:
> First, doing repeated += operations is going to introduce accumulating floating point error.

Totally right.

> Second, I think __iter__ should return self and __next__ should return the next _get_by_index value (meaning, keep track of n like here).

I have a serious doubt here: if `__iter__` returns `self` you have an iterator (that is also an iterable). The issue is that a `numeric_range` becomes stateful, whereas I think it should be stateless (iterable but not iterator) as the regular `range`. Related issues: what is the length of an stateful object? when are stateful objects equal? ...

bbayles replied:
Ah, `next(range(10))` doesn't work - missed that. Anyway, should avoid the floating point error by using a counter, like the current implementation.

jferard replied:
Done.

bbayles commented on more_itertools/more.py:
```diff
@@ -1634,35 +1634,148 @@ def numeric_range(*args):
+            return 0
+
+        r = distance % step
+        q = int(distance / step)
```
Why not `divmod` here?

jferard replied:
Because I initially thought that `5.0 // 1.5` would return `3` and not `3.0`. The code was:

        r = distance % step
        q = distance // step

And then I added a `int` cast and removed the second slash. I'll use `divmod`. Note that there is a subtle difference:

```
>>> divmod(5.0, -1.5)
(-4.0, -1.0)
>>> (5.0//-1.5, 5.0%-1.5)
(-4.0, -1.0)
```

But:
```
>>> (int(5.0/-1.5), 5.0%-1.5)
(-3, -1.0)
```

We do not have this problem here because `distance` and `step` are both positive. I propose an `assert` to document this:

        assert distance > 0 and step > 0

bbayles replied:
Thanks for the explanation. No need for the assert, but could add this as a comment.

jferard replied:
Done.

bbayles commented on more_itertools/more.py:
```diff
@@ -1634,35 +1634,148 @@ def numeric_range(*args):
+            return numeric_range(start, stop, step)
+        else:
+            raise TypeError(
+                "numeric range indices must be integers or slices, not str")
```
`str` is hardcoded here - this should probably check the type of what was passed in, right?

jferard replied:
My mistake!

jferard replied:
Fixed.

bbayles commented on more_itertools/more.py:
```diff
@@ -1654,35 +1654,145 @@ def numeric_range(*args):
+                q, r = divmod(value - self._start, self._step)
+                if r == self._zero:
+                    return int(q)
+        else:
```
I don't think this branch is tested - could you add something that verifies this?

jferard replied:
Done.

bbayles commented on more_itertools/more.py:
```diff
@@ -1654,35 +1654,145 @@ def numeric_range(*args):
+        q, r = divmod(distance, step)
+        return int(q) + int(r != self._zero)
+
+    def __reduce__(self):
```
This method doesn't appear to be tested - could we add a pickle test?

jferard replied:
Done.

jferard commented:
> A couple more requests here, plus one more: could you add some tests of the new methods with Decimal, Fraction, and datetime objects to see what breaks?

I added a lot of tests and fixed some issues. See commit message 37e6b6f for details.

bbayles commented:
Thanks for this!
